pennylane/transforms/qcut.py

```diff
@@ -428,3 +429,320 @@ def contract_tensors(
+                qml.CNOT(wires=[0, 1])
+                qml.expval(qml.PauliZ(1))
+
+            tape_dag = qcut.tape_to_graph(tape)
+
+        Deriving kwargs for a given circuit and feed it to a custom partitioner along with an extra
+        custom parameter:
+        >>> cut_strategy = qcut.CutStrategy(devices=dev)
+        >>> cut_kwargs = cut_strategy.get_cut_kwargs(tape_dag)
+        >>> cut_kwargs.update({'extra_param': 0})
+        >>> my_partitioner(tape_dag, **cut_kwargs)
+
+        Args:
+            tape_dag (MultiDiGraph): Graph representing a tape, typically the output of
+                :func:`tape_to_graph`.
+            max_wires_by_fragment (Sequence[int]): User-predetermined list of wire limits by
+                fragment. If supplied, the number of fragments will be derived from it and
+                exploration of other choices will not be made.
+            max_gates_by_fragment (Sequence[int]): User-predetermined list of gate limits by
+                fragment. If supplied, the number of fragments will be derived from it and
+                exploration of other choices will not be made.
+
+        Returns:
+            List[Dict[str, Any]]: A list of minimal set of kwargs being passed to a graph
+                partitioner method.
+        """
+        tape_wires = set(w for _, _, w in tape_dag.edges.data("wire"))
+        assert all((w is not None for w in tape_wires))
+        num_tape_wires = len(tape_wires)
+        num_tape_gates = tape_dag.order()
+        self._validate_input(max_wires_by_fragment, max_gates_by_fragment)
+
+        probed_cuts = self._infer_probed_cuts(
+            num_tape_wires=num_tape_wires,
+            num_tape_gates=num_tape_gates,
+            max_wires_by_fragment=max_wires_by_fragment,
+            max_gates_by_fragment=max_gates_by_fragment,
+        )
+
+        return probed_cuts
+
+    @staticmethod
+    def _infer_imbalance(
+        k, num_wires, num_gates, free_wires, free_gates, imbalance_tolerance=None
+    ) -> float:
+        """Helper function for determining best imbalance limit."""
+        avg_fragment_wires = (num_wires - 1) // k + 1
+        avg_fragment_gates = (num_gates - 1) // k + 1
+        assert free_wires >= avg_fragment_wires
+        assert free_gates >= avg_fragment_gates
+
+        wire_imbalance = free_wires / avg_fragment_wires - 1
+        gate_imbalance = free_gates / avg_fragment_gates - 1
+        imbalance = min(gate_imbalance, wire_imbalance)
+        if imbalance_tolerance is not None:
+            imbalance = min(imbalance, imbalance_tolerance)
+
+        return imbalance
+
+    @staticmethod
+    def _validate_input(
+        max_wires_by_fragment,
+        max_gates_by_fragment,
+    ):
+        """Helper parameter checker."""
+        if max_wires_by_fragment is not None:
+            assert isinstance(max_wires_by_fragment, (list, tuple))
+            assert all(isinstance(i, int) and i > 0 for i in max_wires_by_fragment)
```

Comment:
Are asserts the best choice for user validation? It's more length by `if-raise` seems more natural.

Reply:
switched.